server/topic.go

```diff
@@ -2956,19 +3007,28 @@ func (t *Topic) addSession(s *Session, asUid types.Uid) bool {
 	return true
 }
 
-// Removes session record if 'asUid' matches subscribed user.
-func (t *Topic) remSession(s *Session, asUid types.Uid) *perSessionData {
-	if pssd, ok := t.sessions[s]; ok && (pssd.uid == asUid || asUid.IsZero()) {
-		// Check for deferred presence notification and cancel it if found.
-		if pssd.ref != nil {
-			t.defrNotif.Remove(pssd.ref)
+// Removes session record if 'doRemove' is true and either:
+// * 'asUid' matches subscribed user.
+// * or 's' is a proxy session for a proxy-master topic connection.
+func (t *Topic) maybeRemoveSession(s *Session, asUid types.Uid, doRemove bool) *perSessionData {
```

Comment:
The logic here is for the use case when the session authenticated with a root user A is subscribed to topic on behalf of user B. It also implements support for deferred online notifications. The `pssd` (per session data) contains user id overrides + reference to deferred notification. 

The session is removed from topic when the actually subscribed user  (A==B) leaves `pssd.uid == asUid` or session is disconnected `asUid.IsZero()` regardless of user (similar to `leave.terminateProxyConnection == true`).

If session is removed from topic then non-nil `pssd` is returned. Otherwise nil is returned.

How it applies to a session from a proxy topic:

* Proxy session at master topic should have zeros and nils in `pssd`.
* `pssd` should not be checked at the master topic (when `s.isProxy` is `true`). `pssd` is for supporting onBehalfOf. This session is always `onBehalfOf` multiple users. It should be checked at proxy topic (`t.isProxy` is `true`).
* The `t.defrNotif.Remove(pssd.ref)` should not be called at the master topic. It should be called  at the origin, at the proxy topic.
* Basically, the proxy session should be unconditionally removed when the origin terminates it. Otherwise it should not be affected.